src/tree/updater_quantile_hist.cc

```diff
@@ -89,39 +88,163 @@ bool QuantileHistMaker::UpdatePredictionCache(
-  double time_build_hist = 0;
-  double time_evaluate_split = 0;
-  double time_apply_split = 0;
+  tstart = dmlc::GetTime();
 
-  const std::vector<GradientPair>& gpair_h = gpair->ConstHostVector();
+  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
+  time_init_data = dmlc::GetTime() - tstart;
 
-  spliteval_->Reset();
+  // FIXME(hcho3): this code is broken when param.num_roots > 1. Please fix it
+  CHECK_EQ(p_tree->param.num_roots, 1) <<
+    "tree_method=hist does not support multiple roots at this moment";
 
+  // in depth_wise growing, we feed loss_chg with 0.0 since it is not used anyway
+  qexpand_depth_wise.push_back(ExpandEntry(0, p_tree->GetDepth(0), 0.0, timestamp++));
+  ++num_leaves;
+  // get stats of node per depth
   tstart = dmlc::GetTime();
-  this->InitData(gmat, gpair_h, *p_fmat, *p_tree);
-  time_init_data = dmlc::GetTime() - tstart;
+  for (int depth = 0; depth < param_.max_depth + 1; depth++) {
+    int starting_index = std::numeric_limits<int>::max();
+    int sync_count = 0;
+    std::vector<ExpandEntry> temp_qexpand_depth;
+    // 0. build histogram for the nodes in current level
+    for (size_t k = 0; k < qexpand_depth_wise.size(); k++) {
+      int nid = qexpand_depth_wise[k].nid;
+      if (rabit::IsDistributed()) {
+        if (IsLeft(nid) || IsRoot(nid)) {
+          // in distributed setting, we always calcuate from left child or root node
+          hist_.AddHistRow(nid);
+          BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+          if (left_to_right_siblings.find(nid) != left_to_right_siblings.end()) {
+            nodes_to_derive[left_to_right_siblings[nid]] = nid;
+          }
+          sync_count++;
+          starting_index = std::min(starting_index, nid);
+        }
+      } else {
+        if (IsLeft(nid) &&
+            (row_set_collection_[nid].Size() <
+             row_set_collection_[left_to_right_siblings[nid]].Size())) {
+          hist_.AddHistRow(nid);
+          BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+          nodes_to_derive[left_to_right_siblings[nid]] = nid;
+          sync_count++;
+          starting_index = std::min(starting_index, nid);
+        } else if (IsRight(nid) && (row_set_collection_[nid].Size() <=
+                                    row_set_collection_[right_to_left_siblings[nid]].Size())) {
+          hist_.AddHistRow(nid);
+          BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+          nodes_to_derive[right_to_left_siblings[nid]] = nid;
+          sync_count++;
+          starting_index = std::min(starting_index, nid);
+        } else if (IsRoot(nid)) {
+          // root node
+          hist_.AddHistRow(nid);
+          BuildHist(gpair_h, row_set_collection_[nid], gmat, gmatb, hist_[nid], false);
+          sync_count++;
+          starting_index = std::min(starting_index, nid);
+        }
+      }
+    }
+    // 1. sync histogram
```

Comment:
If you have to number your comments to describe the flow of this function perhaps these numbered sections should be functions on their own.